v4/export/sql.go

```diff
@@ -1009,3 +1056,154 @@ func GetPartitionNames(db *sql.Conn, schema, table string) (partitions []string,
 	}, "SELECT PARTITION_NAME from INFORMATION_SCHEMA.PARTITIONS WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?", schema, table)
 	return
 }
+
+// GetPartitionTableIDs get partition tableIDs through histograms.
+// SHOW STATS_HISTOGRAMS  has db_name,table_name,partition_name but doesn't have partition id
+// mysql.stats_histograms has partition_id but doesn't have db_name,table_name,partition_name
+// So we combine the results from these two sqls to get partition ids for each table
+// If UPDATE_TIME,DISTINCT_COUNT are equal, we assume these two records can represent one line.
+// If histograms are not accurate or (UPDATE_TIME,DISTINCT_COUNT) has duplicate data, it's still fine.
+// Because the possibility is low and the effect is that we will select more than one regions in one time,
+// this will not affect the correctness of the dumping data and will not affect the memory usage much.
+// This method is tricky, but no better way is found.
+// Because TiDB v3.0.0's information_schema.partition table doesn't have partition name or partition id info
+// return (dbName -> tbName -> partitionName -> partitionID, error)
+func GetPartitionTableIDs(db *sql.Conn, tables map[string]map[string]struct{}) (map[string]map[string]map[string]int64, error) {
+	const (
+		showStatsHistogramsSQL   = "SHOW STATS_HISTOGRAMS"
+		selectStatsHistogramsSQL = "SELECT TABLE_ID,FROM_UNIXTIME(VERSION DIV 262144 DIV 1000,'%Y-%m-%d %H:%i:%s') AS UPDATE_TIME,DISTINCT_COUNT FROM mysql.stats_histograms"
```

Comment:
may need kernel's review about UPDATE_TIME and DISTINCT_COUNT have same value in two queries 🤔

Reply:
@time-and-fate These two values have low possibility to be different.

Reply:
Yes. 
The statistics themselves are stored in the `mysql.stast_xxx` tables. The `SHOW STATS_HISTOGRAMS` display the cache in the TiDB memory.
The results are not always the same with each other, there are 2 main reasons:
(1) If there're changes, there may be inconsistency and it will take several seconds to become consistent again (the time is also configurable).
(2) Statistics on indexed are always loaded. But statistics on columns are loaded by need, they are loaded only after tried to use them.